home/tests.py

```diff
@@ -264,8 +265,37 @@ def test_explore_page_response(self, explore_page_response):
             assert explore_page_response.status_code == 200
             assert explore_page_response.templates[0].name == "home/explore.html"
 
-            questions = Question.objects.all()
-            assert set(explore_page_response.context['questions']) == set(questions)
+        class TestExplorePagePaginator:
+            @pytest.mark.django_db
+            def test_explore_page_is_paginated(self, explore_page_response):
+                assert (explore_page_response.context['is_paginated'] is True)
+
+            @pytest.mark.parametrize("items_per_page", [
+                5,
+                10,
+                15,
+                20
+            ])
+            @pytest.mark.django_db
+            def test_explore_page_items_per_page(self, items_per_page, client):
+                # assert all questions appears in the explore page and corresponding the selected "items_per_page" .
+                questions = list(Question.objects.all())
+
+                last_page = math.ceil(len(questions) / items_per_page)
```

Comment:
Can use the `//` operator for integer division instead of using floating-point functions.

Though in general I would call this test too complex. As soon as you start doing calculations in test code, you introduce the possibility of having bugs in it. Perhaps think of a simpler way to test? Or maybe simplify the feature to make it easier to test? For example, do the users really need to be able to choose the amount of questions they see on the page?

Reply:
the ` //` operator round the result down while I wanted to round it up therefore I used "Math.ceil" function.

Reply:
Math.ceil is a floating-point processing function, you are effectively converting int values to floats, rounding via an expensive FPU operation and then converting back to ints. Doing `//` and `+1` is much cheaper.